metpy/io/pandas_to_netcdf.py

```diff
@@ -0,0 +1,172 @@
+        Type of dataset to be converted. Options are 'timeSeries', 'profile',
+        or 'trajectory'. While optional, this variable should be declared to create
+        a CF-compliant DSG netCDF file.
+
+    Returns
+    -------
+    NetCDF file saved to `path_to_save`.
+
+    """
+    # Verify_integrity must be true in order for conversion to netCDF to work
+    # Return a TypeError if not provided a Pandas DataFrame
+    try:
+        # Create the dimensions for use later in netCDF file
+        samplingindex = df.groupby([sampling_var], sort=False).ngroup()
+        obs = arange(0, len(df))
+        df.insert(0, 'samplingIndex', samplingindex)
+        df.insert(1, 'observations', obs)
+
+        # Handle the sampling location specific data
+        sampling_data = df[sampling_data_vars]
+        samples = sampling_data.groupby([sampling_var], sort=False).ngroup()
+        sampling_data.insert(0, 'samples', samples)
+        sampling_data = sampling_data.groupby('samples').first()
+        dataset_samples = xr.Dataset.from_dataframe(sampling_data)
+
+        # Create the dataset for the variables of each observation
+        df = df.drop(sampling_data_vars, axis=1)
+        df = df.set_index(['observations'], verify_integrity=True)
+        dataset_var = xr.Dataset.from_dataframe(df)
+
+        # Merge the two datasets together
+        dataset_final = xr.merge([dataset_samples, dataset_var], compat='no_conflicts')
+
+    except (AttributeError, ValueError, TypeError):
+        raise TypeError('A pandas dataframe was not provided')
+
+    # Attach variable-specific metadata
+    _assign_metadata(dataset_final, column_units, standard_names, long_names)
+
+    # Attach dataset-specific metadata
+    if dataset_type:
+        dataset_final.attrs['featureType'] = dataset_type
+    else:
+        log.warning('No dataset type provided - netCDF will not have appropriate metadata'
+                    'for a DSG dataset.')
+    if dataset_type:
+        dataset_final[sampling_var].attrs['cf_role'] = dataset_type.lower() + '_id'
+    dataset_final['samplingIndex'].attrs['instance_dimension'] = 'samples'
+
+    # Determine mode to write to netCDF
+    write_mode = 'w'
+    if path.exists(path_to_save):
+        # Eventually switch to 'a' to allow appending and delete error
+        raise ValueError('File already exists - please delete and run again')
+
+    # Check if netCDF4 is installed to see how many unlimited dimensions we can use
+    # Need conditional import for checking due to Python 2
+    try:
+        from importlib.util import find_spec
+        check_netcdf4 = find_spec('netCDF4')
+    except ImportError:
+        from imp import find_module
+        check_netcdf4 = find_module('netCDF4')
+
+    # Make sure path is a string to allow netCDF4 to be used - needed for tests to pass
+    path_to_save = str(path_to_save)
+
+    if check_netcdf4 is not None:
+        unlimited_dimensions = ['samples', 'observations']
```

Comment:
So I've always been told CF doesn't support multiple unlimited dimensions--though I can't find that in the spec anywhere, only many references to "**the** unlimited dimension". @ethanrd @lesserwhirls can you weigh in here?

I'm thinking we may need to (optionally) support the "ragged array" options that are described in the CF spec to collapse the observation and sample dimensions.

Reply:
The single unlimited dimension is a netCDF Classic Model limitation:

> An unlimited dimension has a length that can be expanded at any time, as more data are written to it. NetCDF files can contain at most one unlimited dimension.

but not for the netCDF Enhanced Data Model (netCDF-4):

>  The Enhanced Data Model supports the classic model in a completely backward-compatible way, while allowing access to new features such as groups, multiple unlimited dimensions, and new types, including user-defined types.

See https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_model.html

Now, as far as I understand, CF does not explicitly say which data model is follows (classic vs enhanced), although classic is implied in several places, such as the [discussion on supported data types](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_data_types), examples that only use one unlimited dimension, lack of any mention of the use groups, etc. That does not, however, mean CF only applies to the Classic Data Model. See https://github.com/cf-convention/cf-conventions/issues/191.

So, are multiple unlimited dimensions supported by CF?

![dunno](https://media.giphy.com/media/13Qumr2SLqrl5e/giphy.gif)

Reply:
Glad I wasn't the only one confused by the CF specs...

Reply:
I can change this back to one unlimited dimension only if desired. My thought for two dimensions is if we get this to appending mode in the future, we would need to append along the "station" and "time" dimensions for METARs for example. Right now, given that we aren't including the ability to append due to Xarray's limitations, we could just go with one unlimited dimension and update later when needed/if the CF specs are clarified at any point.